### ISSUE TITLE
I2S switches from MSB-first to LSB-first on an ESP32-S3

### ISSUE THREAD

**Issue by ma261065** [micropython/micropython#11245]
Using the latest nightly build (v1.19.1-1014-gbde222ce8 (2023-04-11) ) on an ESP32-S3 with SPIRAM, I am sending a wav file that is stored on the onboard filesystem to a PCM5102A DAC via I2S.

When I run the code the first time after power on or reset (soft or hard) the audio is heard clearly.

If I then run the same code again, it is just a mess of static - I can *just* hear the beat of the music in all the noise though.

The only way to get it working again is to reboot. This happens on I2S0 as well as I2S1, and have tried different pins in case I was clashing with one of the ESP32 pins that has alternative functions.

Also, when rebooting I see the following message:
_E (130390) I2S: i2s_driver_uninstall(2047): I2S port 0 has not installed_
just before the _MPY: soft reboot_ message

I am not getting this problem on a non-S3 variant.

Code below:

    from machine import I2S, Pin
    import io

    sck_pin = Pin(13)   # Serial clock output
    ws_pin = Pin(14)    # Word clock output
    sd_pin = Pin(17)    # Serial data output

    buf = bytearray(1024)

    audio_out = I2S(0, sck=sck_pin, ws=ws_pin, sd=sd_pin, mode=I2S.TX, bits=16, format=I2S.STEREO, rate=16000, ibuf=20000)

    with io.open("example.wav", 'rb') as fp:
        size = fp.readinto(buf)
 
        while (size > 0):
            print('.', end='')
            audio_out.write(buf)
            size = fp.readinto(buf)
        
    print("Finished")    
    fp.close()
    audio_out.deinit()

**Comment by ma261065** [comment]
From the ESP-IDF docs, I note that the `i2s_std_slot_config_t` struct for the ESP32-S3 is different to a normal ESP32, in that the ESP32-S3 one contains an extra two bools at the end, and changes the meaning of the last bool:

**ESP32**
...
...
bool bit_shift - Set to enable bit shift in Philips mode
bool msb_right - Set to place right channel data at the MSB in the FIFO

vs

**ESP32-S3**
...
...
bool bit_shift - Set to enable bit shift in Philips mode
bool left_align - Set to enable left alignment
bool big_endian - Set to enable big endian
bool bit_order_lsb - Set to enable lsb first

So I guess if we are passing the ESP32 `i2s_std_slot_config_t` struct to an ESP32-S3, then the last two bools will be off in random memory somewhere and could lead to inconsistent behaviour. And given that one of them controls MSB-first/LSB-first behaviour, this could explain the issue. Am I right?

**Comment by ma261065** [comment]
OK - I think I have it figured out.

In `esp-idf/components/driver/deprecated/driver/i2s_types_legacy.h` we have the following IF-delimited extra section in the `i2s_config_t` struct, if the chip supports TDM:

```
...
...
#if SOC_I2S_SUPPORTS_TDM
    i2s_channel_t           chan_mask;                  /*!< I2S active channel bit mask, set value in i2s_channel_t to enable specific channel, the bit map of active channel can not exceed (0x1<<total_chan). */
    uint32_t                total_chan;                 /*!< I2S Total number of channels. If it is smaller than the biggest active channel number, it will be set to this number automatically. */
    bool                    left_align;                 /*!< Set to enable left alignment */
    bool                    big_edin;                   /*!< Set to enable big endian */
    bool                    bit_order_msb;              /*!< Set to enable msb order */
    bool                    skip_msk;                   /*!< Set to enable skip mask. If it is enabled, only the data of the enabled channels will be sent, otherwise all data stored in DMA TX buffer will be sent */
#endif // SOC_I2S_SUPPORTS_TDM
```

At this point the ESP32C3, ESP32C6 & ESP32S3 support TDM

I had always thought that uninitialised members of a struct in C got set to default values, but it seems that this isn't happening here**.



In `machine_i2s.c` if I add the following to the end where `machine_i2s_init_helper` populates the `i2s_config` struct, then it all works reliably, and doesn't change behaviour between executions.

```
#if SOC_I2S_SUPPORTS_TDM
    i2s_config.chan_mask = 0x03;
    i2s_config.total_chan = 0;
    i2s_config.left_align = true;
    i2s_config.big_edin = false;
    i2s_config.bit_order_msb = false;
    i2s_config.skip_msk = false;
#endif
```

** I can't explain this, but if I put debug printf statements in `machine_i2s_init_helper`, those bools always show as true even though `machine_i2s_init_helper` doesn't set them, validating my thinking that they are being set to default values. However the behaviour still varies between executions, even though the printf always shows the same. If I explicitly set them as per the above, the random behaviour disappears.

**Comment by MATTYGILO** [comment]
Hi I am having the same problem, has this been merged into micropython yet?

**Comment by MATTYGILO** [comment]
> OK - I think I have it figured out.
> 
> In `esp-idf/components/driver/deprecated/driver/i2s_types_legacy.h` we have the following IF-delimited extra section in the `i2s_config_t` struct, if the chip supports TDM:
> 
> ```
> ...
> ...
> #if SOC_I2S_SUPPORTS_TDM
>     i2s_channel_t           chan_mask;                  /*!< I2S active channel bit mask, set value in i2s_channel_t to enable specific channel, the bit map of active channel can not exceed (0x1<<total_chan). */
>     uint32_t                total_chan;                 /*!< I2S Total number of channels. If it is smaller than the biggest active channel number, it will be set to this number automatically. */
>     bool                    left_align;                 /*!< Set to enable left alignment */
>     bool                    big_edin;                   /*!< Set to enable big endian */
>     bool                    bit_order_msb;              /*!< Set to enable msb order */
>     bool                    skip_msk;                   /*!< Set to enable skip mask. If it is enabled, only the data of the enabled channels will be sent, otherwise all data stored in DMA TX buffer will be sent */
> #endif // SOC_I2S_SUPPORTS_TDM
> ```
> 
> Currently the ESP32C3, ESP32C6 & ESP32S3 support TDM
> 
> I had always thought that uninitialised members of a struct in C got set to default values, but it seems that this isn't happening here**.
> 
> In `machine_i2s.c` if I add the following to the end where `machine_i2s_init_helper` populates the `i2s_config` struct, then it all works reliably, and doesn't change behaviour between executions.
> 
> ```
> #if SOC_I2S_SUPPORTS_TDM
>     i2s_config.chan_mask = 0x03;
>     i2s_config.total_chan = 0;
>     i2s_config.left_align = true;
>     i2s_config.big_edin = false;
>     i2s_config.bit_order_msb = false;
>     i2s_config.skip_msk = false;
> #endif
> ```
> 
> ** I can't explain this, but if I put debug printf statements in `machine_i2s_init_helper`, those bools always show as true even though `machine_i2s_init_helper` doesn't set them, validating my thinking that they are being set to default values. However the behaviour still varies between executions, even though the printf always shows the same. If I explicitly set them as per the above, the random behaviour disappears.

Can confirm this works

**Comment by dpgeorge** [comment]
This may be fixed by 0b145fdc2d4857a77f9f0a3662646d52da61d5b2

**Comment by ma261065** [comment]
> This may be fixed by [0b145fd](https://github.com/micropython/micropython/commit/0b145fdc2d4857a77f9f0a3662646d52da61d5b2)

Confirmed that this update to the new I2S driver fixes the issue.

**Comment by ma261065** [comment]
Fixed by [0b145fd](https://github.com/micropython/micropython/commit/0b145fdc2d4857a77f9f0a3662646d52da61d5b2)
